mongo_connector/dest_mapping.py

```diff
@@ -15,143 +15,230 @@
+        return RegexSet(regexes, strings)
+
+
+class DestMapping(object):
+    """Manages included and excluded namespaces.
+    """
+    def __init__(self, namespace_set=None, ex_namespace_set=None,
+                 user_mapping=None, include_fields=None,
+                 exclude_fields=None):
         # a dict containing plain mappings
         self.plain = {}
-        # a dict containing wildcard mappings
-        self.wildcard = {}
+        # a list of (RegexObject, MappedNamespace) tuples
+        self.regex_map = []
         # a dict containing reverse plain mapping
         # because the mappings are not duplicated,
         # so the values should also be unique
         self.reverse_plain = {}
         # a dict containing plain db mappings, db -> a set of mapped db
         self.plain_db = {}
 
+        # Fields to include or exclude from all namespaces
+        self.include_fields = include_fields
+        self.exclude_fields = exclude_fields
+
         # the input namespace_set and ex_namespace_set could contain wildcard
-        self.namespace_set = namespace_set
-        self.ex_namespace_set = ex_namespace_set
+        self.namespace_set = set()
+        self.ex_namespace_set = RegexSet.from_namespaces(
+            ex_namespace_set or [])
 
         self.lock = threading.Lock()
 
+        user_mapping = user_mapping or {}
+        namespace_set = namespace_set or []
         # initialize
         for ns in namespace_set:
             user_mapping.setdefault(ns, ns)
 
-        for k, v in user_mapping.items():
-            self.set(k, v)
-
-    def set_plain(self, key, value):
-        """A utility function to set the corresponding plain variables"""
-        if value in self.reverse_plain:
-            raise errors.InvalidConfiguration(
-                "Destination namespaces set should not"
-                " contain any duplicates.")
-
-        db, col = key.split(".", 1)
-        self.plain[key] = value
-        self.reverse_plain[value] = key
-        if col != "$cmd":
-            if db not in self.plain_db:
-                self.plain_db[db] = set([value.split(".")[0]])
+        for src_name, v in user_mapping.items():
+            if isinstance(v, dict):
+                self._add_mapping(src_name, v.get('rename'))
             else:
-                self.plain_db[db].add(value.split(".")[0])
-
-    def match(self, src, dst_pattern):
-        """If source string src matches dst, return the matchobject"""
-        reg_pattern = r'\A' + dst_pattern.replace('*', '(.*)') + r'\Z'
-        m = re.match(reg_pattern, src)
-        return m
+                self._add_mapping(src_name, v)
 
-    def match_set(self, plain_src, dst_arr):
-        for x in dst_arr:
-            if plain_src == x:
-                return True
-            if "*" in x:
-                m = self.match(plain_src, x)
-                if m:
-                    return True
+    def _add_mapping(self, src_name, dest_name=None, include_fields=None,
```

Comment:
This might better be called `_add_collection` or `_add_collection_spec` since this is dealing with more than just the mapping now.